master/internal/core_experiment.go

```diff
@@ -356,43 +363,41 @@ type CreateExperimentParams struct {
 func (m *Master) parseCreateExperiment(params *CreateExperimentParams) (
 	*model.Experiment, bool, *tasks.TaskSpec, error,
 ) {
-	resources := model.ParseJustResources([]byte(params.ConfigBytes))
-	taskSpec := m.makeTaskSpec(resources.ResourcePool, resources.SlotsPerTrial)
-
-	config := model.DefaultExperimentConfig(&taskSpec.TaskContainerDefaults)
-
-	checkpointStorage, err := m.config.CheckpointStorage.ToModel()
+	// Read the config as the user provided it.
+	config, err := expconf.ParseAnyExperimentConfigYAML([]byte(params.ConfigBytes))
 	if err != nil {
 		return nil, false, nil, errors.Wrap(err, "invalid experiment configuration")
 	}
 
-	config.CheckpointStorage = *checkpointStorage
-
+	// Apply the template that the user specified.
 	if params.Template != nil {
 		template, terr := m.db.TemplateByName(*params.Template)
 		if terr != nil {
 			return nil, false, nil, terr
 		}
-		if yerr := yaml.Unmarshal(template.Config, &config, yaml.DisallowUnknownFields); yerr != nil {
+		var tc expconf.ExperimentConfig
+		if yerr := yaml.Unmarshal(template.Config, &tc, yaml.DisallowUnknownFields); yerr != nil {
 			return nil, false, nil, yerr
 		}
+		// Merge the template into the config.
+		config = schemas.Merge(config, tc).(expconf.ExperimentConfig)
 	}
 
-	if yerr := yaml.Unmarshal(
-		[]byte(params.ConfigBytes), &config, yaml.DisallowUnknownFields,
-	); yerr != nil {
-		return nil, false, nil, errors.Wrap(yerr, "invalid experiment configuration")
-	}
+	// Merge the appropriate TaskContainerDefaults into the config.
+	resources := schemas.WithDefaults(config).(expconf.ExperimentConfig).Resources()
+	taskSpec := m.makeTaskSpec(resources.ResourcePool(), resources.SlotsPerTrial())
+	taskSpec.TaskContainerDefaults.MergeIntoConfig(&config)
 
-	if config.Environment.PodSpec == nil {
-		if config.Resources.SlotsPerTrial == 0 {
-			config.Environment.PodSpec = m.config.TaskContainerDefaults.CPUPodSpec
-		} else {
-			config.Environment.PodSpec = m.config.TaskContainerDefaults.GPUPodSpec
-		}
-	}
+	// Merge in the master's checkpoint storage into the config.
+	config.RawCheckpointStorage = schemas.Merge(
+		config.RawCheckpointStorage, &m.config.CheckpointStorage,
+	).(*expconf.CheckpointStorageConfig)
+
+	// Lastly, apply any json-schema-defined defaults.
+	config = schemas.WithDefaults(config).(expconf.ExperimentConfig)
 
-	err = check.Validate(config)
+	// Make sure the experiment config has all eventuallyRequired fields.
+	err = schemas.IsComplete(config)
 	if err != nil {
```

Comment:
```suggestion
	if err = schemas.IsComplete(config); err != nil {
```
just a nit. when my brain parses go code, as is, `err` goes into memory until it goes out of scope, where like suggested it is gc'd immediately.

Reply:
In general I prefer a pattern which is repeatable everywhere over a pattern which is not.  So I get really frustrated when you do `x, err := ...` and `x` isn't defined after the if statement.  It seems like a really, really pointless optimization.

Especially when the `err` object is never appended to or anything crazy so it never matters if you have one `err` instance in a function or a bunch of them.  Also, `err` is always nil, or you exit the function right away anyway, so I doubt there's ever a material difference in GC behavior.

I've been accused of not being a very idiomatic go coder though.

Reply:
> In general I prefer a pattern which is repeatable everywhere over a pattern which is not.

ad absurdum, there is no pattern which is repeatable everywhere, there are different patterns for different things. one is a pattern when the non of the variables are used later in the scope, the other is for when they are.. it's just a mental optimization for reading code faster for me, and I notice it does make a slight different since my brain is extra finite.

> so I doubt there's ever a material difference in GC behavior.

when i was talking about gc behavior, i meant my _mental_ garbage collector lol. when i see
```
err = func()
if err != nil {
  ...
}
...
```
my brain holds onto err. when i see 
```
if err = func(); err != nil {
  ...
}
it doesn't. This is a micro-optimization for readability to me. Maybe I read code worse/slower than others so i actually notice it lol.

Reply:
i'm really just explaining why i even care to comment this. sorry.